Update ScrollView after notification closes in Editor Layers View

jesusbotella opened the pull request:
The code in this PR fixes ScrollView not being updated in Editor's Layers View when notification closes after adding a layer, leaving an unexpected scrollbar where scrolling is not possible because of a wrong height.

We need to trigger a ScrollView update because the notification DOM node is not a child of the mentioned ScrollView, so perfect-scrollbar does not know that a layout change has been made.

#### Changes proposed in this pull request:

- [editor/editor-pane.js]()
  - Inject `updateScrollWhenExist` function from ScrollView to be available in LayersView instance
- [editor/layers/layers-view.js]()
  - Add `updateScrollWhenExist` to LayerView required options
  - Listen to `notification:close` event
  - Trigger ScrollView update on `notification:close` event handler

#### Current behaviour after fix

![fix](https://user-images.githubusercontent.com/4319728/32691504-a690deae-c708-11e7-8242-769280651bb5.gif)

#### Questions
- Do you want me to add the test to a new describe section in order not to do Notifier setup and teardown before and after each test and do it only in the notification test?

**Fixes**: #12953

Cartofante commented:
Can one of the admins verify this patch?

Cartofante commented:
Can one of the admins verify this patch?

ivanmalagon commented on lib/assets/core/javascripts/cartodb3/editor/editor-pane.js:
```diff
@@ -106,7 +106,8 @@ module.exports = CoreView.extend({
                       visDefinitionModel: self._visDefinitionModel,
-                      showMaxLayerError: self._infoboxState.bind(self)
+                      showMaxLayerError: self._infoboxState.bind(self),
+                      updateScrollWhenExist: this._updateScrollWhenExist.bind(this)
```
The rest of parameters uses `self` instead of `this`. I guess we ran into problems here before. If you have any good reason not to use `self` instead of `this`, please tell us. In other case, change it and check that it still works.

matallo replied:
probably `onNotificationCloseAction` or similar would be a better parameter name in the same fashion than other actions passed as arguments in this same view, i.e. `onClickAction`

https://github.com/jesusbotella/cartodb/blob/2f28ecbc22b1ec3c9cf68a2ae6059026da412513/lib/assets/core/javascripts/cartodb3/editor/editor-pane.js#L118

jesusbotella commented:
I have just updated the PR with the changes.

@matallo You are right. I didn't notice that variable below. Consistency is king.  :)

@ivanmalagon I'm afraid it wouldn't work using `self` (I tested it) because `this` in that scope refers to the ScrollView instance. The variable assigned to `self` is the EditorPane instance [in line 62](https://github.com/jesusbotella/cartodb/blob/3fe9294b26e3f117e59a12a8c935970647094cd2/lib/assets/core/javascripts/cartodb3/editor/editor-pane.js#L62), and the EditorPane doesn't have any method named `updateScrollWhenExist`. 

But I agree with you that it is tricky for anyone to read, so I think that it would be better to assign `this` to a variable named `scrollViewInstance`, so anyone that reads it know what it is happening there. I have updated the PR with that change; please let me know what you think.

I tried to make it easier when I thought the best way to achieve the ScrollView update, but I didn't find any more readable and straightforward way to inject the function from the ScrollView to the LayersView :(

ramiroaznar commented:
Why this was not applied?

jesusbotella commented:
It was the code test that I did prior joining CARTO 😂

ramiroaznar commented:
Can we merge it?

Jesus89 commented:
I'm taking this one

Jesus89 commented:
It is working OK (broken CI because of the external fork)

![scroll-ok](https://user-images.githubusercontent.com/2227572/47161851-24bc3480-d2f3-11e8-9a92-ed24fd48943d.gif)
